[Docs] Add Sandboxes page (#9772)
skypilot-org/skypilot: docs/source/sandboxes.rst

Beyond launch / exec / down, the ``sky_sandbox`` SDK covers batch and async
fan-out, templates and warm pools, and secret / volume injection:

.. tab-set::

    .. tab-item:: Batch

        Pass ``num_sandboxes`` to launch a batch in one call; names are
        auto-generated from the prefix (``rollout-0001``, ``rollout-0002``, ...).

        .. code-block:: python

            import sky_sandbox

            sandboxes = sky_sandbox.launch(
                name='rollout',
                template='ml-gpu',
                num_sandboxes=1000,
            )
            for sb in sandboxes:
                sb.exec('python rollout.py')

    .. tab-item:: Async fan-out

        A single event loop can drive hundreds of concurrent ``exec`` calls.

        .. code-block:: python

            import asyncio
            import sky_sandbox

            async def main():
                sandboxes = await sky_sandbox.launch_async(
                    name='rollout', num_sandboxes=100)
                results = await asyncio.gather(
                    *(sb.exec_async('python rollout.py') for sb in sandboxes))
                await asyncio.gather(*(sb.down_async() for sb in sandboxes))

>                 await sky_sandbox.aclose()  # release the shared session

Author: Good catch, thanks. Done in b393500af: the async fan-out example now wraps the execs in `try...finally` so the sandboxes are always torn down and the shared session closed even if an `exec` raises.

.. _skypilot-sandboxes:

.. rst-class:: hero-title

> Sandboxes

Reviewer: Let's add a new badge for this page

Author: Done in 1533c77db: added the **New** badge to the Sandboxes sidebar entry, and the **For Frontier AI** section now auto-expands so it is visible.

Kubernetes clusters. Each sandbox is a lightweight pod you can launch on demand,
run commands in, and tear down, without provisioning a full cluster.

Sandboxes are built for **AI coding agents**, **RL training rollouts**, and
**parallel evals**: workloads that need many short-lived, isolated environments
spun up and down quickly. Pre-warmed pools launch sandboxes in **under a
second**, with volumes and secrets injected automatically. A built-in image for
`Claude Code <https://www.anthropic.com/claude-code>`_ ships out of the box.

.. tip::

   Sandboxes are part of **SkyPilot Platform**, in limited early access.
   `Sign up here <https://forms.gle/HGGMjzvRz8Mqn9pn7>`_; takes 20 seconds.

.. raw:: html

   <figure class="align-center" style="width: 90%; margin: 0 auto 20px auto;">
     <video id="sandbox-video" style="width: 100%; height: auto;" autoplay muted playsinline loop>
        <source src="_static/sandbox-claude.mp4" type="video/mp4" />
     </video>
     <figcaption><p>Sandboxes: launching a Claude Code environment on your own cluster with secrets injected from the SkyPilot Secrets Manager.</p></figcaption>
   </figure>

> Why sandboxes?

Reviewer: how about: "Why sandboxes with SkyPilot"

Author: Done in 1533c77db: renamed to **Why sandboxes with SkyPilot**.

  from other agents.
* **RL training rollouts:** spin up thousands of sandboxes to run rollouts in
  parallel, collect results, and tear them down, all from a single process.
* **Parallel evals:** run a large eval suite across many isolated environments at
  once instead of serializing on one machine.
* **Ephemeral build and CI tasks:** execute short-lived builds, tests, or scripts
  in a clean environment without provisioning a full cluster.

How it works
------------

A **template** defines the shape of a sandbox: its container image, CPU, memory,
and any baked-in volumes. Each template is backed by a **warm pool** of idle
pods, so launches against it are instant. SkyPilot ships a built-in ``default``
template (a ``python`` image); create your own when you need a different image or
size.

A **sandbox** is a single running pod claimed from a template's warm pool. It has
a unique name, a lifecycle you control (launch, use, then tear down), and is
accessible until you call ``down()``. Commands run inside the sandbox over the
same exec primitive ``kubectl exec`` uses, returning ``stdout``, ``stderr``, and

> the exit code synchronously.

Reviewer: ```suggestion
accessible until you call ``down()``.
```

Author: The SDK renamed `down` to `terminate` (Modal-parity) in the latest version, but applied otherwise.

> 

Reviewer: Possible to have some diagram to show how template and sandbox work together? Also, the quick start does not include the template. Can we show quick start first, and make template an advanced concept?

Author: Done in 1533c77db: added an embedded animated diagram of how templates, warm pools, and sandboxes fit together, moved Quickstart ahead of any template concepts, and split templates into their own **Templates and warm pools** section.

Because each sandbox is its own pod, it is isolated from other sandboxes and
workloads on the cluster, a clean boundary for code an agent generates at
runtime. Secrets referenced at launch are resolved by the SkyPilot Secrets
Manager and injected as environment variables, so they are never stored in the

> template image or passed on the command line.

Reviewer: Do we need this paragraph?

Author: Removed in 1533c77db.

Quickstart
----------

Once the SkyPilot CLI and the bundled Sandbox SDK are installed, launch a
sandbox, run a command, and tear it down:

.. code-block:: python

>    import sky_sandbox

Reviewer: Is it `sky_sandbox`? Can it be part of `sky.sandbox`?

Author: Done: the page now uses `sky.sandbox` throughout (13692541e).

       sb.exec('python train.py')
   # Sandbox is torn down on exit.

Working with the SDK
--------------------

Beyond launch / exec / down, the ``sky_sandbox`` SDK covers batch and async
fan-out, templates and warm pools, and secret / volume injection:

.. tab-set::

    .. tab-item:: Batch

        Pass ``num_sandboxes`` to launch a batch in one call; names are
        auto-generated from the prefix (``rollout-0001``, ``rollout-0002``, ...).

        .. code-block:: python

            import sky_sandbox

            sandboxes = sky_sandbox.launch(
                name='rollout',
                template='ml-gpu',
                num_sandboxes=1000,
            )
            for sb in sandboxes:
                sb.exec('python rollout.py')

    .. tab-item:: Async fan-out

        A single event loop can drive hundreds of concurrent ``exec`` calls.

        .. code-block:: python

            import asyncio
            import sky_sandbox

            async def main():
                sandboxes = await sky_sandbox.launch_async(
                    name='rollout', num_sandboxes=100)
                try:
                    results = await asyncio.gather(

>                         *(sb.exec_async('python rollout.py') for sb in sandboxes))

Reviewer: Might be more interesting if it is different per-sandbox
```suggestion
                    results = await asyncio.gather(
                        *(sb.exec_async(f'python rollout.py {i}') for i, sb in enumerate(sandboxes)))
```

Author: Done in 1533c77db: the batch and async examples now pass a per-sandbox arg via `enumerate`.

                finally:
                    # Always tear down, even if an exec raises.
                    await asyncio.gather(*(sb.down_async() for sb in sandboxes))
                    await sky_sandbox.aclose()  # release the shared session

            asyncio.run(main())

>     .. tab-item:: Templates & warm pools

Reviewer: This should be in the advanced section for templates / warm pools

Author: Done in 1533c77db: moved into the new **Templates and warm pools** section.

            import sky_sandbox

            # Create a template with 10 warm pods kept idle and ready.
            sky_sandbox.create_template(
                name='ml-gpu',
                image='nvcr.io/nvidia/pytorch:24.05-py3',
                cpus=8,
                memory_gb=64,
                pool_replicas=10,
            )

            # Scale the warm pool up or down at any time.
            sky_sandbox.set_pool_size('ml-gpu', replicas=50)

    .. tab-item:: Secrets & volumes

        Inject secrets and mount volumes at launch.

        .. code-block:: python

            import sky_sandbox

            sb = sky_sandbox.launch(
                name='job',
                template='ml-gpu',
                # Inject a stored secret as an env var (via the secrets manager).
                secret_refs=[{'name': 'HF_TOKEN', 'env_var_name': 'HF_TOKEN'}],
                # Plain (non-secret) env vars.
                env={'PROJECT': 'demo'},
                # Mount an existing volume (create it with `sky volumes apply`).
                volume_mounts=[{'volume_name': 'shared-data',
                                'mount_path': '/data'}],
            )

>     .. tab-item:: Claude Code

Reviewer: This seems a very different use case. Could we make it a different section instead of a tab?

Author: Done in 1533c77db: promoted to its own **Running AI coding agents** section.

        The built-in ``claude`` template ships with Claude Code installed; pass
        an OAuth token via the secrets manager and drive it non-interactively.

        .. code-block:: python

            import sky_sandbox

            with sky_sandbox.launch(
                name='claude',
                template='claude',
                secret_refs=[{'name': 'CLAUDE_CODE_OAUTH_TOKEN'}],
            ) as sb:
                result = sb.exec(
                    "echo 'Create a hello world index.html' | "
                    'claude -p --dangerously-skip-permissions')
                print(result['stdout'])

The SDK exposes ``launch``, ``exec``, ``down``, ``ls``, ``create_template``,
``set_pool_size``, and ``delete_template``, each with an ``_async`` variant for
event-loop code. See the dashboard's **Sandboxes** page to manage templates,
warm pools, and running sandboxes in the UI.

.. seealso::

   - :ref:`volumes-all`: persistent storage you can mount into sandboxes.

>    - :ref:`job-groups`: run many parallel jobs and sandboxes together for RL.

Reviewer: Do we have a reference in those section or how do they use sandboxes?

Author: We talked about this; we will add references for how those features use sandboxes in a follow-up.

* **Secrets stay out of your code:** credentials are injected at launch from the
  SkyPilot Secrets Manager as environment variables, so tokens are never baked
  into images or hardcoded into the commands an agent runs.
* **Massively parallel:** launch thousands of sandboxes in one call for RL
  rollouts and parallel evals, then fan out commands concurrently.
* **Runs on your infra:** sandboxes live on your own Kubernetes clusters, so your
  code and data never leave your environment, and capacity is simply your
  existing cluster.

Use cases
---------

* **AI coding agents:** give each agent (e.g. Claude Code) its own disposable
  environment to read, write, and run code in, isolated from your other work and
  from other agents.
* **RL training rollouts:** spin up thousands of sandboxes to run rollouts in
  parallel, collect results, and tear them down, all from a single process.
* **Parallel evals:** run a large eval suite across many isolated environments at
  once instead of serializing on one machine.
* **Ephemeral build and CI tasks:** execute short-lived builds, tests, or scripts
  in a clean environment without provisioning a full cluster.

Quickstart
----------

> A sandbox is a pod you create, run commands in, and terminate. Once the SkyPilot

Reviewer: ```suggestion
A sandbox is a isolated pod you create, run commands in, and terminate. Once the SkyPilot
```

                    await sky.sandbox.aclose()  # release the shared session

            asyncio.run(main())

    .. tab-item:: Secrets & volumes

        Inject secrets and mount volumes at create time.

        .. code-block:: python

            import sky.sandbox

            sb = sky.sandbox.create(
                name='job',
                # Inject secrets from the secrets manager as env vars of the
                # same name.
                secrets=['HF_TOKEN'],
                # Plain (non-secret) env vars.
                env={'PROJECT': 'demo'},
                # Mount existing volumes, keyed by mount path (create them with
                # `sky volumes apply`).
                volumes={'/data': 'shared-data'},
            )

The SDK exposes ``create``, ``exec``, ``terminate``, ``ls``, ``create_template``,
``set_pool_size``, and ``delete_template``. Every per-call entrypoint has an async
sibling on a ``.aio`` attribute (``sky.sandbox.create.aio(...)``,
``sb.exec.aio(...)``), so the same names work in event-loop code. See the
dashboard's **Sandboxes** page to manage templates, warm pools, and running

> sandboxes in the UI.

Reviewer: Not urgent, but we should do a SDK reference at the end than a list of SDK functions here.

Author: Sounds good!

> Running AI coding agents

Reviewer: ```suggestion
Example: Running AI coding agents
```

------------------------

The built-in ``claude`` template ships with `Claude Code
<https://www.anthropic.com/claude-code>`_ installed. Give an agent its own
isolated sandbox, inject its token from the secrets manager, and drive it
non-interactively:

.. code-block:: python

   import sky.sandbox

   with sky.sandbox.create(
       name='claude',
       template='claude',
       secrets=['CLAUDE_CODE_OAUTH_TOKEN'],
   ) as sb:
       # Shell features like pipes need an explicit shell.
       result = sb.exec(
           'sh', '-c',
           "echo 'Create a hello world index.html' | "
           'claude -p --dangerously-skip-permissions')
       print(result['stdout'])

Each agent runs in its own pod, so many agents can work in parallel without
sharing a filesystem or stepping on each other.

> Templates and warm pools

Reviewer: ```suggestion
Advanced: Templates and warm pools for fast provisioning
```

* **Secrets stay out of your code:** credentials are injected at launch from the
  SkyPilot Secrets Manager as environment variables, so tokens are never baked
  into images or hardcoded into the commands an agent runs.
* **Massively parallel:** launch thousands of sandboxes in one call for RL
  rollouts and parallel evals, then fan out commands concurrently.
* **Runs on your infra:** sandboxes live on your own Kubernetes clusters, so your
  code and data never leave your environment, and capacity is simply your
  existing cluster.

Use cases
---------

* **AI coding agents:** give each agent (e.g. Claude Code) its own disposable
  environment to read, write, and run code in, isolated from your other work and
  from other agents.
* **RL training rollouts:** spin up thousands of sandboxes to run rollouts in
  parallel, collect results, and tear them down, all from a single process.
* **Parallel evals:** run a large eval suite across many isolated environments at
  once instead of serializing on one machine.
* **Ephemeral build and CI tasks:** execute short-lived builds, tests, or scripts
  in a clean environment without provisioning a full cluster.

Quickstart
----------

> 

Reviewer: Should we include the CLI way somewhere which is shown in the video?

Author: Good call, done. The Quickstart is now a CLI / Python tab-set: the CLI tab shows `sky sandbox create` (drops into a shell, destroyed on exit), plus the detached `create` / `ls` / `terminate` flow.

The SDK exposes ``create``, ``exec``, ``terminate``, ``ls``, ``create_pool``,
``set_pool_size``, and ``delete_pool``. Every per-call entrypoint has an async
sibling on a ``.aio`` attribute (``sky.sandbox.create.aio(...)``,
``sb.exec.aio(...)``), so the same names work in event-loop code. See the
dashboard's **Sandboxes** page to manage pools and running sandboxes in the UI.

Running AI coding agents
------------------------

The built-in ``claude`` pool ships with `Claude Code
<https://www.anthropic.com/claude-code>`_ installed. Give an agent its own
isolated sandbox, inject its token from the secrets manager, and drive it
non-interactively:

.. code-block:: python

   import sky.sandbox

   with sky.sandbox.create(
       name='claude',
       pool='claude',
       secrets=['CLAUDE_CODE_OAUTH_TOKEN'],
   ) as sb:
       # Shell features like pipes need an explicit shell.
       result = sb.exec(
           'sh', '-c',
           "echo 'Create a hello world index.html' | "
           'claude -p --dangerously-skip-permissions')
       print(result['stdout'])

Each agent runs in its own pod, so many agents can work in parallel without
sharing a filesystem or stepping on each other.

> Pools

Reviewer: ```suggestion
Advanced: Warm pools
```

-----

> A **pool** defines the shape of a sandbox (its container image, CPU, and memory)

Reviewer: We should mention why people may want to use pool, e.g., how much faster etc?

Reviewer: By reading this doc, I have no idea why I need a pool.

         var s = doc.createElement('style');
         s.id = 'sky-embed-transparent';
         s.textContent =
           'div[style*="margin: 0px auto"][style*="border-radius: 16px"]{background:transparent !important;border:none !important;box-shadow:none !important;}' +
           'div[style*="margin: 0px auto"][style*="border-radius: 16px"] > div{background:transparent !important;}' +
           'div[style*="radial-gradient"]{background:transparent !important;background-image:none !important;}';
         doc.head.appendChild(s);
       } catch (e) { /* cross-document access can fail harmlessly; ignore */ }
     }
     iframe.addEventListener('load', strip);
     strip();
   })();
   </script>

Create a pool, and resize it at any time:

.. code-block:: python

   import sky.sandbox

   # Create a pool with 10 warm pods kept idle and ready.
   sky.sandbox.create_pool(
       name='ml-gpu',
       image='nvcr.io/nvidia/pytorch:24.05-py3',
       cpus=8,
       memory_gb=64,
       replicas=10,
   )

   # Scale the pool up or down at any time.
   sky.sandbox.set_pool_size('ml-gpu', replicas=50)

   # Launch a sandbox from the pool.

>    sb = sky.sandbox.create(name='train', pool='ml-gpu')

Reviewer: not urgent, but we should make the following work:
```
   # Create a pool with 10 warm pods kept idle and ready.
   pool = sky.sandbox.create_pool(
       name='ml-gpu',
       image='nvcr.io/nvidia/pytorch:24.05-py3',
       cpus=8,
       memory_gb=64,
       replicas=10,
   )

   # Scale the pool up or down at any time.
   pool.resize(replicas=50)

   # Launch a sandbox from the pool.
   sb = sky.sandbox.create(name='train', pool=pool)
```

Author: Agreed!
